### PR TITLE
parsing failure context label

### LINKED ISSUE #95: parsing failure context label
It would be nice if we could get some context with parse failures, like Megaparsec [`label`](https://hackage.haskell.org/package/megaparsec-9.0.0/docs/Text-Megaparsec.html#v:label)

In `ParserT`, `fail` is implemented as [`throwError`](https://pursuit.purescript.org/packages/purescript-transformers/4.2.0/docs/Control.Monad.Error.Class#t:MonadError)

https://github.com/purescript-contrib/purescript-parsing/blob/e801a0ef42f3211b1602a94a269eef7ce551423f/src/Text/Parsing/Parser.purs#L132-L138

This suggests a cheap implementation strategy for `label`, using [`catchError :: forall a. m a -> (e -> m a) -> m a`](https://pursuit.purescript.org/packages/purescript-transformers/4.2.0/docs/Control.Monad.Error.Class#v:catchError).

```purescript
-- | If parsing fails inside this context, then prepend the String to the error String.
label :: forall m s a. Monad m => String -> ParserT s m a -> ParserT s m a
label preerror p = catchError p go
 where go (ParseError err pos) = failWithPosition (preerror <> err) pos
```

Actually come to think of it, anyone can use this `label` function without library support. Maybe I'll just do that.

I suppose if we wanted to make this more official and support it in the library, then we could change the error type to `ParseError (NonEmptyList String) Position` and cons the errors up as we fail out of the parser.

### PR DESCRIPTION
**Pull request by jamesdbrock**
## What does this pull request do?

Resolves #95 

Also, in a separate commit, add a link in the README to __purescript-parsing-dataview__, as offered by @garyb https://github.com/purescript-contrib/purescript-parsing/issues/88#issuecomment-589978391

## Where should the reviewer start?

There is one function added, and one test.

## How should this be manually tested?

`pulp --psc-package test`

### CODE REVIEW

File: src/Text/Parsing/Parser.purs
Comment (thomashoneyman) [inline src/Text/Parsing/Parser.purs:144]: Is there value in appending with a space by default, or should that be left to the user?

Comment (jamesdbrock) [reply src/Text/Parsing/Parser.purs:144]: Yeah good question. I say leave it to the user, who might reasonably choose to append `" "` or `"\n"` or something else.

**Review by thomashoneyman — APPROVED** [review]

**Comment by jamesdbrock** [comment]
This is actually more like a Megaparsec [`region`](https://hackage.haskell.org/package/megaparsec-9.0.0/docs/Text-Megaparsec.html#v:region) than a [`label`](https://hackage.haskell.org/package/megaparsec-9.0.0/docs/Text-Megaparsec.html#v:label). Maybe the name of the function should be different.

**Comment by thomashoneyman** [comment]
I don't have an opinion on which of those names is used for the function, but another maintainer may. If you think that `region` is more informative, then I am OK with that.

**Comment by jamesdbrock** [comment]
Oh now I remember why I thought the name `label` would be good, because this is the same function as
[`label`](https://hackage.haskell.org/package/binary-0.8.8.0/docs/Data-Binary-Get.html#v:label) from the `Data.Binary.Get` parsing monad.

**Comment by thomashoneyman** [comment]
I'm going to go ahead and merge this, but if you would like to revisit the name prior to the next release of the library then please feel free to open an issue / PR. Thanks!

**Comment by jamesdbrock** [comment]
Cool I think it's good the way it is, thank you @thomashoneyman

**Comment by jamesdbrock** [comment]
I just realized that this should be

```purescript
label :: forall m s a. Monad m => (ParseError -> ParseError) -> ParserT s m a -> ParserT s m a
```

So that the user can do whatever they want to the `ParseError`, not just prepending a `String`.

Also probably better for performance, since the `String` argument will get strictly evaluated, but this `(ParseError -> ParseError)` funciton would only be called on failure.

**Comment by jamesdbrock** [comment]
Oh hi! Ok, I'll make the update.

**Pull request merged**
